sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ShiftDynamicNotIntegerBase.cs

```diff
@@ -0,0 +1,83 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class ShiftDynamicNotIntegerBase<TExpressionSyntax> : SonarDiagnosticAnalyzer
+        where TExpressionSyntax : SyntaxNode
+    {
+        internal const string DiagnosticId = "S3449";
+        protected const string MessageFormat = "Remove this erroneous shift, it will fail because '{0}' can't be implicitly converted to 'int'.";
+
+        protected abstract DiagnosticDescriptor Rule { get; }
```

Comment:
I don't like using `SupportedDiagnostics[0]` because it implies that there could be more than one supported diagnostics and is weird why we chose the first to report and not the second... I know it is used in some rules, but in general I prefer the more explicit single `Rule` override, which is also used in many rules.

Reply:
But reporting with `Rule` doesn't prevent to register more than ` rule in the base class collection.